tfx/dsl/experimental/spans_resolver.py

```diff
@@ -0,0 +1,117 @@
+  ) -> base_resolver.ResolveResult:
+    artifacts_dict = {}
+    resolve_state_dict = {}
+    pipeline_context = metadata_handler.get_pipeline_context(pipeline_info)
+    if pipeline_context is None:
+      raise RuntimeError('Pipeline context absent for %s' % pipeline_context)
+    for k, c in source_channels.items():
+      if c.type_name != Examples.TYPE_NAME:
+        raise ValueError('Resolving non-Example artifacts is not supported.')
+
+      candidate_artifacts = metadata_handler.get_qualified_artifacts(
+          contexts=[pipeline_context],
+          type_name=c.type_name,
+          producer_component_id=c.producer_component_id,
+          output_key=c.output_key)
+
+      previous_artifacts = sorted(
+          candidate_artifacts, 
+          key=lambda a: int(
+              a.artifact.custom_properties[utils.SPAN_PROPERTY_NAME].string_value),
+          reverse=True)
+
+      artifacts_dict[k] = []
+      resolve_state_dict[k] = False
+
+      if self._range_config:
+        if self._range_config.HasField('static_range'):
+          # TODO(jjma): Optimize this by sending a more specific MLMD query.
+          lower_bound = self._range_config.static_range.start_span_number
+          upper_bound = self._range_config.static_range.end_span_number
+
+          for a in previous_artifacts:
+            span = int(
+                a.artifact.custom_properties[utils.SPAN_PROPERTY_NAME].string_value)
+            if lower_bound <= span and span <= upper_bound:
+              artifacts_dict[k].append(
+                  artifact_utils.deserialize_artifact(a.type, a.artifact))
+        
+          resolve_state_dict[k] = (
+              len(artifacts_dict[k]) == (upper_bound - lower_bound + 1))
+
+        elif self._range_config.HasField('rolling_range'):
+          start_span = self._range_config.rolling_range.start_span_number
+          num_spans = self._range_config.rolling_range.num_spans
+          num_skip = self._range_config.rolling_range.skip_num_recent_spans
+
+          for i, a in enumerate(previous_artifacts[num_skip:]):
+            span = int(
+                a.artifact.custom_properties[utils.SPAN_PROPERTY_NAME].string_value)
+            if span >= start_span and i < num_spans:
+              artifacts_dict[k].append(
+                  artifact_utils.deserialize_artifact(a.type, a.artifact))
+            else:
+              break
+          
+          resolve_state_dict[k] = len(artifacts_dict[k]) == num_spans
+
+      elif len(previous_artifacts) > 0:
+        # Default behavior is to fetch single latest span artifact.
+        latest_artifact = previous_artifacts[0]
+        artifacts_dict[k] = [
+            artifact_utils.deserialize_artifact(
+                latest_artifact.type, latest_artifact.artifact)
+        ]
+        resolve_state_dict[k] = True
+
+    return base_resolver.ResolveResult(
+        per_key_resolve_result=artifacts_dict,
+        per_key_resolve_state=resolve_state_dict)
```

Comment:
I'm thinking even we have multiple examplegen as input source, but we only want to output a single example artifact channel for trainer

e.g., a for loop of examplegens that process span 3,4,5 and a resolver that resolves a channel of example artifact (3,4,5).

the artifacts_dict[k], can k be the type instead of name?

_another option (need to verify if it works or not) is using [this](https://github.com/tensorflow/tfx/blob/master/tfx/components/base/base_node.py#L153) for resolver to add examplegen dependency, and resolver just resolve based on component type (not id) and artifact type_

Reply:
i see what you mean (same as the paste in the doc right?). the design for this resolver is more following the logic for the LatestArtifactResolver, where its possible for a resolver to resolve multiple channels. are we fixing this resolver to only resolver one channel always? if so, we can definitely make k be the type name and instead of creating an output for each source channel, we can just aggregate them all into one output.

Reply:
for our use case, we still want the artifact comes from certain component (e.g., example artifact comes from examplegens, not transforms), 

if we just have one examplegen in the pipeline, current code would be fine.

but if we config multiple examplegen in the pipeline, we need a way to get a channel for all examplegen's output artifacts

maybe add a boolean merge_same_artifact_type or sth, if set to true, same type will be treated as a single (the 'a' below, not 'b')

a.multiple channel with same type -> merge -> process -> output single channel
b.multiple channel with same type -> process -> merge -> output single channel


_for the other option I mentioned above, channel need to contains the component type info besides component id info_

Reply:
updated PR to merge multiple channels if flag is set. it merges these into a resolved output with key thats the same as the type name `Examples`.

Reply:
Span is a special concept that only applies for `Examples`. Combining multiple `Channel`s into one can be handled separately. Shall we just assert that this resolver takes exactly one source which is of type `Examples`? This will also make migration to the soon-to-be-there new stack easier. This will also simplify the logic a bit.

Reply:
i think from a design standpoint it might make more sense. are you proposing having a 2nd resolver that can resolve multiple channels of input into a single channel of output?

Reply:
from discussion with hui and ruoyu today, this resolver should mainly be for taking one input channel (of Examples) and retrieving the range of spans. likely a different primitive will handle the case where we want to resolve multiple example_gens into a single channel, so to make the logic cleaner, we will simply force this resolver to only conssider one channel.